File: Example/ObjCDemo/ObjCDemo/DetailViewController.m
//
//  DetailViewController.m
//  ObjCDemo
//
//  Created by LoginRadius Development Team on 18/05/16.
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

#import "DetailViewController.h"
#import "LoginRadius.h"
#import "XLFormViewControllerExtension.h"


static NSArray<NSString *>* _countries = nil;
@interface DetailViewController ()
@property (strong, nonatomic) NSString *accessToken;
@property (strong, nonatomic) NSDictionary *userProfile;
@end

@implementation DetailViewController

+ (NSArray<NSString *>*)countries {
    if (!_countries) {
        NSLocale *locale = [NSLocale currentLocale];
        NSArray *countryArray = [NSLocale ISOCountryCodes];

        NSMutableArray<NSString *> *countryArr = [[NSMutableArray<NSString *> alloc] init];

        for (NSString *countryCode in countryArray) {

            NSString *displayNameString = [locale displayNameForKey:NSLocaleCountryCode value:countryCode];
            [countryArr addObject:displayNameString];

        }

        [countryArr sortUsingSelector:@selector(localizedCompare:)];
        _countries = [countryArr copy];
    }

    return _countries;
}

    //NSArray<NSString *> *countries = NSLocale

      /* NSLocale.isoCountryCodes.map { (code:String) -> String in
        let id = NSLocale.localeIdentifier(fromComponents: [NSLocale.Key.countryCode.rawValue: code])
        return NSLocale(localeIdentifier: "en_US").displayName(forKey: NSLocale.Key.identifier, value: id) ?? "Country not found for code: \(code)"
    }*/

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
     [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(setupForm)
                                                 name:UIApplicationWillEnterForegroundNotification
                                               object:nil];
    
    [self setupForm];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillEnterForegroundNotification object:nil];
}

- (void) dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)setupForm{
    NSString * userAccessToken =  [[[LoginRadiusSDK sharedInstance] session] accessToken];
    if (userAccessToken == nil)
    {
        [self showAlert:@"ERROR" message:@"Access token is missing or logged out"];
        [self logoutPressed:self];
        return;
    }
    
    self.accessToken = userAccessToken;
    [self validateAccessToken:NO];
    
    NSDictionary * profile =  [[[LoginRadiusSDK sharedInstance] session] userProfile];
    
    if(profile)
    {
        if ([self userProfile] && [profile isEqualToDictionary:[self userProfile]])
        {
            //same userProfile, don't reload ui
            return;
        }
        
        self.userProfile = profile;
    }else{
        profile = [[NSDictionary alloc] init];
    }
    
    [[self navigationItem] setHidesBackButton:YES animated:NO];
    [self navigationItem].title = @"User Profile ObjC";
    
    XLFormDescriptor * form;
    XLFormSectionDescriptor * section;
    XLFormRowDescriptor * row;
    XLFormRowDescriptor * switchRow;
    
    form = [XLFormDescriptor formDescriptor];
    
    section = [XLFormSectionDescriptor formSection];
    [form addFormSection:section];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Logout" rowType:XLFormRowDescriptorTypeButton title:@"Log out"];
    row.action.formBlock = ^(XLFormRowDescriptor *block)
    {
        [self logoutPressed:self];
    };
    [section addFormRow:row];
    NSString *userEmail;
    if ([_userProfile objectForKey:@"Email"] != nil && ([[_userProfile objectForKey:@"Email"] count] > 0)) {
     userEmail = [[_userProfile objectForKey:@"Email"][0] objectForKey:@"Value"];
    }
    //User Profile Section
    
    NSString *userCountry;
    if ([_userProfile objectForKey:@"Country"] != nil && ![[_userProfile objectForKey:@"Country"]  isEqual: @""]) {
         userCountry = [[_userProfile objectForKey:@"Country"] objectForKey:@"Name"];
    }
    NSString *gender = [_userProfile objectForKey:@"Gender"];

    section = [XLFormSectionDescriptor formSectionWithTitle:@"User Profile Section"];
    [form addFormSection:section];
    
    switchRow = [XLFormRowDescriptor formRowDescriptorWithTag:@"UserProfile" rowType:XLFormRowDescriptorTypeBooleanSwitch title:@"User Profile"];
    switchRow.value = @1;
    [section addFormRow:switchRow];

    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"FirstName" rowType:XLFormRowDescriptorTypeText title:@"First Name"];
    row.required = YES;
    //bug on validator returning nil when theres no validator
    [row addValidator:[XLFormRegexValidator formRegexValidatorWithMsg:@"First Name is required!" regex:@".+"]];
    row.value = [_userProfile objectForKey:@"FirstName"];
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    [row.cellConfig setObject:@(NSTextAlignmentRight) forKey:@"textField.textAlignment"];
    [section addFormRow:row];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"LastName" rowType:XLFormRowDescriptorTypeText title:@"Last Name"];
    row.required = YES;
    [row addValidator:[XLFormRegexValidator formRegexValidatorWithMsg:@"Last Name is required!" regex:@".+"]];
    row.value = [_userProfile objectForKey:@"LastName"];
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    [row.cellConfig setObject:@(NSTextAlignmentRight) forKey:@"textField.textAlignment"];
    [section addFormRow:row];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Email" rowType:XLFormRowDescriptorTypeEmail title:@"Email"];
    row.required = YES;
    row.value = userEmail;
    row.disabled = @YES;
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    [row.cellConfig setObject:@(NSTextAlignmentRight) forKey:@"textField.textAlignment"];
    [section addFormRow:row];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Gender" rowType:XLFormRowDescriptorTypeSelectorSegmentedControl title:@"Gender"];
    row.selectorOptions = @[@"M",@"F",@"?"];
    row.value = gender;
    [row addValidator:[XLFormRegexValidator formRegexValidatorWithMsg:@"Gender is required!" regex:@".+"]];
    row.required = YES;
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    [section addFormRow:row];
    
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Country" rowType:XLFormRowDescriptorTypeSelectorPush title:@"Country"];
    row.selectorOptions = [DetailViewController countries];
    row.value = userCountry;
    [row addValidator:[XLFormRegexValidator formRegexValidatorWithMsg:@"Country is required!" regex:@".+"]];
    row.required = YES;
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    [section addFormRow:row];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Update Profile" rowType:XLFormRowDescriptorTypeButton title:@"Update Profile"];
    row.hidden = [NSString stringWithFormat:@"$%@ == 0", switchRow];
    row.action.formSelector = @selector(validateUserProfileInput);
    [section addFormRow:row];
    
    section = [XLFormSectionDescriptor formSection];
    [form addFormSection:section];
    
    row = [XLFormRowDescriptor formRowDescriptorWithTag:@"Information" rowType:XLFormRowDescriptorTypeButton title:@"Information"];
    row.action.formBlock = ^(XLFormRowDescriptor *block)
    {
        [self showAlert:@"Information" message:@"For more usage examples,\nSee our Swift demo!"];
    };
    [section addFormRow:row];

    self.form = form;
}

- (void)validateUserProfileInput {
    XLFormRowDescriptor *fNameRow = [[self form] formRowWithTag:@"FirstName"];
    XLFormRowDescriptor *lNameRow = [[self form] formRowWithTag:@"LastName"];
    XLFormRowDescriptor *genderRow = [[self form] formRowWithTag:@"Gender"];
    XLFormRowDescriptor *countryRow = [[self form] formRowWithTag:@"Country"];
    
    NSMutableArray<XLFormValidationStatus *> *errors = [[NSMutableArray<XLFormValidationStatus *> alloc] init];
    
    [errors addObject:[fNameRow doValidation]];
    [errors addObject:[lNameRow doValidation]];
    [errors addObject:[genderRow doValidation]];
    [errors addObject:[countryRow doValidation]];

    for (int i = 0; i < [errors count]; i++)
    {
        if (!errors[i].isValid)
        {
            [self showAlert:@"ERROR" message:[errors[i] msg]];
            return;
        }
    }
    
    [self updateProfile:@[fNameRow,lNameRow,genderRow,countryRow]];
}

- (void)updateProfile: (NSArray<XLFormRowDescriptor *> *) rows
{
    NSMutableDictionary *parameters = [[NSMutableDictionary alloc]init];
    
    for(int i = 0; i < [rows count]; i++)
    {
        XLFormRowDescriptor *row = rows[i];
        if ([[row tag]  isEqual: @"Country"]) {
            NSDictionary *country = @{@"Code":@"",@"Name":[row value]};
            [parameters setObject:country forKey:@"Country"];
        }else{
            [parameters setObject:[row value] forKey:[row tag]];
        }
    }

   
    
    [[AuthenticationAPI authInstance] updateProfileWithAccessToken:[self accessToken] emailtemplate:nil smstemplate:nil payload:[parameters copy] completionHandler:^(NSDictionary *data, NSError *error) {
        if (error)
        {
            [self showAlert:@"ERROR" message:[error localizedDescription]];
        }
        else{
                    NSString *access_token =  [[[LoginRadiusSDK sharedInstance] session] accessToken];
                    [self showAlert:@"SUCCESS" message:@"User updated!"];
                    NSLog(@"Here is the raw NSDictionary user profile:");
                    NSDictionary *_data =[data objectForKey:@"Data"];
                    LRSession *session = [[LRSession alloc] initWithAccessToken:access_token userProfile:[[_data mutableCopy] replaceNullWithBlank]];
                    NSLog(@"LRSession Store Update Profile%@",session.userProfile);
                    NSLog(@"end of raw NSDictionary user profile");
                    [self viewDidLoad];
                }
        
        
    }];
}

- (void) validateAccessToken:(BOOL)showSuccess
{
    

    [[AuthenticationAPI authInstance] validateAccessToken:[self accessToken] completionHandler:^(NSDictionary * data, NSError *error) {
        if (error)
        {
            [self showAlert:@"ERROR" message:[error localizedDescription]];
        }else if (showSuccess)
        {
            NSString *msg = [NSString stringWithFormat:@"Access Token: %@ is VALID\n expires in: %@",[data objectForKey:@"access_token"], [data objectForKey:@"expires_in"]];
            [self showAlert:@"SUCCESS" message: msg];
        }
    }];
}

- (void)logoutPressed:(id)sender {
    /* Google Native Sign in
    [[GIDSignIn sharedInstance] signOut];
    */
    /* Twitter Native Sign in
    [self twitterLogout];
    */

    [LoginRadiusSDK logout];
    [self.navigationController popViewControllerAnimated:YES];
}

/* Twitter Native Sign in
- (void) twitterLogout
{
    NSArray * twitterSessions;
    twitterSessions = [[[Twitter sharedInstance] sessionStore] existingUserSessions];
    if (twitterSessions){
        for (id session in twitterSessions){
            [[[Twitter sharedInstance] sessionStore] logOutUserID:[session userID]];
        }
    }
}
*/

@end

File: Example/SwiftDemo/SwiftDemo/DetailViewController.swift
//
//  DetailViewController.swift
//  SwiftDemo
//
//  Created by LoginRadius Development Team on 19/05/16.
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

import Eureka
import SwiftyJSON
import LoginRadiusSDK
/* Google Native SignIn
import GoogleSignIn
*/

/* Twitter Native Sign in
import TwitterKit
*/

class DetailViewController: FormViewController {
    
    //List of countries provided from Apple's NSLocale class
    let countries = NSLocale.isoCountryCodes.map { (code:String) -> String in
        let id = NSLocale.localeIdentifier(fromComponents: [NSLocale.Key.countryCode.rawValue: code])
        return NSLocale(localeIdentifier: "en_US").displayName(forKey: NSLocale.Key.identifier, value: id) ?? "Country not found for code: \(code)"
    }
    
    var accessToken:String!
    var userProfile:JSON!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        //This is do when there are 2 apps sharing 1 LoginRadius sitename login
        //If the other app logged out, this logs out too.
        NotificationCenter.default.addObserver(self, selector: #selector(self.setupForm), name:  UIApplication.willEnterForegroundNotification, object: nil)
        
        setupForm()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self, name: UIApplication.willEnterForegroundNotification, object: nil)

    }
    
    deinit
    {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc func setupForm(){
        guard let userAccessToken = LoginRadiusSDK.sharedInstance().session.accessToken
        else
        {
            self.showAlert(title: "ERROR", message: "Access token is missing or logged out")
            self.logoutPressed()
            return
        }
        
        accessToken = userAccessToken
        validateAccessToken(showSuccess: false)
        
        if let userDict = LoginRadiusSDK.sharedInstance().session.userProfile
        {
            if let oldProfile = self.userProfile,
                oldProfile == JSON(userDict)
            {
                //same userProfile, don't reload ui
                return
            }
        
            userProfile = JSON(userDict)
        }else{
            userProfile = JSON([])
        }
        
        //Small UI modification
        self.navigationItem.setHidesBackButton(true, animated: false)
        self.navigationItem.title = "User Profile Swift"

        //Form setup
        self.form = Form()
        
        self.form  +++ Section("")
        <<< ButtonRow("Log out")
        {
            $0.title = $0.tag
            }.onCellSelection{ cell, row in
                self.logoutPressed()
        }
        
        self.smallUserProfileUISetup()
        self.accessTokenUISetup()
        self.customObjectUISetup()
        
    }
    
    // User Profile API Demo
    
    /// Validates simple profile input, see the form construction for form rules
    func validateUserProfileInput()
    {
        let fName = form.rowBy(tag: "FirstName")!
        let lName = form.rowBy(tag: "LastName")!
        let gender = form.rowBy(tag: "Gender")!
        let country = form.rowBy(tag: "Country")!
    
        var errors = fName.validate()
        errors += lName.validate()
        errors += gender.validate()
        errors += country.validate()

        if errors.count > 0{
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        updateProfile(rows:[fName,lName,gender,country])
    }
    
    func updateProfile(rows:[BaseRow])
    {
        var parameters:[String:Any] = [:]
        
        for row in rows
        {
            parameters[row.tag!] = row.baseValue
        
        }
        let country = parameters["Country"] as! String
        let Gender = parameters["Gender"] as! String
        let FirstName = parameters["FirstName"] as! String
        let LastName = parameters["LastName"] as! String
        
        let Country:AnyObject = ["Code":"Primary",
                               "Name":country
            ] as AnyObject
       
        let par:AnyObject = [ "Country":Country,
           "Gender": Gender,
           "FirstName": FirstName,
           "LastName": LastName
            ]as AnyObject

       
    
        AuthenticationAPI.authInstance().updateProfile(withAccessToken:accessToken, emailtemplate: "", smstemplate: "", payload: par as! [AnyHashable : Any], completionHandler: { (data, error) in
            
            if let err = error
            {
                print(err.localizedDescription)
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else
            {
                let access_token = LoginRadiusSDK.sharedInstance().session.accessToken
                self.showAlert(title: "SUCCESS", message: "User updated!")
                print("Here is the raw NSDictionary user profile:")
                let session = LRSession.init(accessToken:access_token as! String, userProfile:data!["Data"] as! [AnyHashable : Any])
                print(data as Any)
                print("end of raw NSDictionary user profile")
                self.viewDidLoad()
            }
        
        })
    }

    // End of User Profile API Demo
    
    // Access Token API Demo
    
    func validateAccessToken(showSuccess:Bool = true)
    {
        
        
        
        AuthenticationAPI.authInstance().validateAccessToken(accessToken, completionHandler: {(data, error) in
            
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else if let d = data,
                    let at = d["access_token"],
                    let expiry = d["expires_in"],
                    showSuccess
            {
                self.showAlert(title: "SUCCESS", message: "Access Token: \(at) is VALID\n expires in: \(expiry)")
            }
        })
    }
    
    func invalidateAccessToken()
    {
       
        AuthenticationAPI.authInstance().invalidateAccessToken(accessToken, completionHandler: {(data, error) in
            
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else if let _ = data
            {
                self.showAlert(title: "SUCCESS", message: "Access Token now is invalid Please Click on Logout Button And try to Login Again")
            }
        })
    }
    
    // Custom Object API Demo
    func createCustomObject()
    {
        var errors = form.rowBy(tag: "CrCO Object Name")!.validate()
        errors += form.rowBy(tag: "CrCO Data")!.validate()

        if errors.count > 0
        {
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        let objectName = form.rowBy(tag: "CrCO Object Name")!.baseValue! as! String
        let dataString = form.rowBy(tag: "CrCO Data")!.baseValue! as! String
        let dataJSON = JSON(parseJSON: dataString)
        guard let dataDict = dataJSON.dictionaryObject else
        {
            showAlert(title: "ERROR", message: "Invalid JSON payload")
            return
        }
        
     
    
        CustomObjectAPI.customInstance().createCustomObject(withObjectName: objectName, accessToken: accessToken, payload: dataDict, completionHandler: {data, error in
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else{
                let prettyJson = JSON(data as Any)
                self.showAlert(title: "SUCCESS", message: prettyJson.rawString() ?? "")
            }
        })
    }
    
    func getCustomObject()
    {
        var errors = form.rowBy(tag: "GetCO Object Name")!.validate()

        if errors.count > 0
        {
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        let objectName = form.rowBy(tag: "GetCO Object Name")!.baseValue! as! String
        

        CustomObjectAPI.customInstance().getCustomObject(withObjectName: objectName, accessToken: accessToken, completionHandler: {data, error in
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else{
                let prettyJson = JSON(data as Any)
                self.showAlert(title: "SUCCESS", message: prettyJson.rawString() ?? "nil data")
            }
        })
    }
    
            
    func getCustomObjectWithRecordId()
    {
        var errors = form.rowBy(tag: "GetCO Object Name with RecordId")!.validate()
        errors = form.rowBy(tag: "GetCO RecordId with RecordId")!.validate()

        if errors.count > 0
        {
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        let objectName = form.rowBy(tag: "GetCO Object Name with RecordId")!.baseValue! as! String
        let recordId = form.rowBy(tag: "GetCO RecordId with RecordId")!.baseValue! as! String
        
        CustomObjectAPI.customInstance().getCustomObject(withObjectRecordId:recordId, accessToken:accessToken, objectname:objectName, completionHandler: {data, error in
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else{
                let prettyJson = JSON(data as Any)
                self.showAlert(title: "SUCCESS", message: prettyJson.rawString() ?? "nil data")
            }
        })
    }
    
    func putCustomObject()
    {
        var errors = form.rowBy(tag: "PutCO Object Name")!.validate()
        errors += form.rowBy(tag: "PutCO RecordId")!.validate()
        errors += form.rowBy(tag: "PutCO Data")!.validate()

        if errors.count > 0
        {
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        let objectName = form.rowBy(tag: "PutCO Object Name")!.baseValue! as! String
        let recordId = form.rowBy(tag: "PutCO RecordId")!.baseValue! as! String
        let dataString = form.rowBy(tag: "PutCO Data")!.baseValue! as! String
        let dataJSON = JSON(parseJSON: dataString)
        
        guard let dataDict = dataJSON.dictionaryObject else
        {
            showAlert(title: "ERROR", message: "Invalid JSON payload")
            return
        }
        

        CustomObjectAPI.customInstance().updateCustomObject(withObjectName: objectName, accessToken: accessToken, objectRecordId:recordId, updatetype:"replace", payload: dataDict, completionHandler: {data, error in
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else{
                let prettyJson = JSON(data as Any)
                self.showAlert(title: "SUCCESS", message: prettyJson.rawString() ?? "You are successfully Update the specified Custom Object")
            }
        })
    }
    
    func delCustomObject()
    {
        var errors = form.rowBy(tag: "DelCO Object Name")!.validate()
        errors += form.rowBy(tag: "DelCO RecordId")!.validate()

        if errors.count > 0
        {
            showAlert(title: "ERROR", message: errors[0].msg)
            return
        }
        
        let objectName = form.rowBy(tag: "DelCO Object Name")!.baseValue! as! String
        let recordId = form.rowBy(tag: "DelCO RecordId")!.baseValue! as! String
        
        CustomObjectAPI.customInstance().deleteCustomObject(withObjectRecordId:recordId, accessToken:accessToken, objectname:objectName, completionHandler: {data, error in
            if let err = error
            {
                self.showAlert(title: "ERROR", message: err.localizedDescription)
            }else{
                let prettyJson = JSON(data as Any)
                self.showAlert(title: "SUCCESS", message:prettyJson.rawString() ?? "You are successfully Delete the specified Custom Object")
            }
        })
    }
    
    // end of Custom Object API
    
    // UI Functions
    
    func showFullProfileController ()
    {
        self.performSegue(withIdentifier: "full profile", sender: self);
    }
    
    func logoutPressed()
    {
        /* Google Native Sign in
        GIDSignIn.sharedInstance().signOut()
        */

        /* Twitter Native Sign in
        twitterLogout()
        */
        LoginRadiusSDK.logout()
        let _ = self.navigationController?.popViewController(animated: true)
    }
    
    /* Twitter Native Sign in
    func twitterLogout(){
        if let twitterSessions = Twitter.sharedInstance().sessionStore.existingUserSessions() as? [TWTRAuthSession]{
            for session in twitterSessions{
                Twitter.sharedInstance().sessionStore.logOutUserID(session.userID)
            }
        }
    }
    */
    
    func toggleRedBorderShowErrorMessage(cell:UIView, row:BaseRow)
    {
        if !row.isValid {
            cell.layer.borderWidth = 3
            cell.layer.borderColor = UIColor.red.cgColor
        }else{
            cell.layer.borderWidth = 0
            cell.layer.borderColor = UIColor.clear.cgColor
        }
    }
}
